src/zap/zapinfo.cpp

```diff
@@ -1203,11 +1191,32 @@ void ZapInfo::setEHinfo(unsigned EHnumber,
 
             resolveToken(&resolvedToken);
 
-            CORINFO_CLASS_HANDLE systemObjectHandle = getBuiltinClass(CLASSID_SYSTEM_OBJECT);
-            _ASSERTE(systemObjectHandle == resolvedToken.hClass);
-#endif // _DEBUG
+            if (m_zapper->m_pOpt->m_compilerFlags.IsSet(CORJIT_FLAGS::CORJIT_FLAG_IL_STUB))
+            {
+                // IL stub tokens are 'private' and do not resolve correctly in their parent module's metadata.
 
-            ilClause->ClassToken = mdTypeRefNil; 
+                // Currently, the only place we are using a token here is for a COM-to-CLR exception-to-HRESULT
+                // mapping catch clause.  We want this catch clause to catch all exceptions, so we override the
+                // token to be mdTypeRefNil, which used by the EH system to mean catch(...)
+#ifdef _DEBUG
+                // The proper way to do this, should we ever want to support arbitrary types here, is to "pre-
+                // resolve" the token and store the TypeHandle in the clause.  But this requires additional
+                // infrastructure to ensure the TypeHandle is saved and fixed-up properly.  For now, we will
+                // simply assert that the original type was System.Object.
+
+                CORINFO_CLASS_HANDLE systemObjectHandle = getBuiltinClass(CLASSID_SYSTEM_OBJECT);
+                _ASSERTE(systemObjectHandle == resolvedToken.hClass);
+#endif // _DEBUG
+                ilClause->ClassToken = mdTypeRefNil;
+            }
+            else if (m_currentMethodModule != m_pImage->m_hModule)
```

Comment:
Why would be want to do this for `if (m_currentMethodModule != m_pImage->m_hModule)` only? I think this condition is always false for the regular R2R compilation.

Reply:
This was meant to be a different check. I wanted to check for the module of the catch clause class instead. Thanks for spotting it.

Reply:
I've added this wrong check after testing the change.

Reply:
I do not think we want to check the module of the catch clause either. The catch clause type in current module may depend on types in other modules - those may fail to load as well.

Reply:
I've fixed it, it should be correct now.

Reply:
Ah, I've missed your previous comment. Hmm, ok, so we need to generate the fixup unconditionally.